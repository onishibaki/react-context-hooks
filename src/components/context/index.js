import { createContext, useContext } from "react";

export const EmployeesContext = createContext();

export const useEmployeeState = () => useContext(EmployeesContext);
