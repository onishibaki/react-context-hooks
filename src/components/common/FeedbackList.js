import React from "react";
import { useEmployeeState } from "../context";
import styled from "styled-components";

import { ActionTypes, updateEmployeeFlag } from "../action";
import { Feedback } from "./Feedback";

const FeedbackListContainer = styled.div`
  background: rgba(0, 0, 0, 0.5);
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  width: 100%;
  height: 100%;
  text-align: center;
`;

const FeedbackListWrapper = styled.div`
  border: 1px solid rgb(17, 17, 17);
  background-color: white;
  display: inline-block;
  position: relative;
`;

const Button = styled.button`
  text-decoration: none;
  border: 1px solid #000000;
  padding: 10px 20px;
  font-weight: 700;
  color: black;
  font-size: 12px;
  margin: 25px 10px 10px 10px;
  display: inline-block;
  height: 100%;
  &:hover {
    background: #00000059;
    color: white;
  }
`;

const Unordered = styled.ul`
  list-style: none;
  padding: 0;
  display: flex;
  justify-content: space-around;
  flex-direction: column;
  padding: 0 20px;
  li {
    display: flex;
    margin: 5px 0;
  }
`;

const Paragraph = styled.p`
  border: 1px solid black;
  padding: 5px;
  margin-right: 20px;
  text-align: left;
  word-wrap: break-word;
  font-size: 12px;
  width: 500px;
  height: 54px;
`;

const SpanButton = styled.span`
  font-weight: 700;
  color: black;
  font-size: 15px;
  margin: 10px 0;
  display: inline-block;
  position: absolute;
  top: -8px;
  right: 9px;
  cursor: pointer;
}
`;

export const FeedBackList = () => {
  const [
    {
      form,
      employee: { employee_id, feedbacklist },
    },
    dispatch,
  ] = useEmployeeState();

  const feedBackParam = {
    employeeId: employee_id,
    flag: 1,
  };

  return (
    form === "view" && (
      <FeedbackListContainer className="feedback-list-container">
        <FeedbackListWrapper>
          <SpanButton
            onClick={() =>
              dispatch({
                type: ActionTypes.SET_FORM_STATE,
                form: "",
                data: {},
              })
            }
          >
            X
          </SpanButton>
          <Button
            onClick={() => {
              dispatch({
                type: ActionTypes.SET_FORM_STATE_FEEDBACK,
                form: "save",
                data: {},
                id: employee_id,
              });
            }}
          >
            Add Feedback
          </Button>
          <Button
            onClick={() => {
              updateEmployeeFlag(feedBackParam);
              dispatch({
                type: ActionTypes.SET_FORM_STATE,
                form: "",
                data: {},
              });
            }}
          >
            Assign Feedback
          </Button>
          <Unordered>
            {feedbacklist.map((feedbacklists) => {
              const { feedback_id, feedback } = feedbacklists;
              return feedback_id !== null ? (
                <li key={feedback_id}>
                  <Paragraph>{feedback}</Paragraph>
                  <Button
                    onClick={() => {
                      dispatch({
                        type: ActionTypes.SET_FORM_STATE_FEEDBACK,
                        form: "update",
                        data: feedbacklists,
                        id: employee_id,
                      });
                    }}
                  >
                    Update
                  </Button>
                </li>
              ) : (
                <h5 key="0">No Feedback</h5>
              );
            })}
          </Unordered>
          <Feedback />
        </FeedbackListWrapper>
      </FeedbackListContainer>
    )
  );
};
