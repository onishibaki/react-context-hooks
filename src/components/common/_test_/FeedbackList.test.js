import React from "react";
import { render, cleanup } from "@testing-library/react";
import renderer from "react-test-renderer";
import Adapter from "enzyme-adapter-react-16";
import { configure, mount } from "enzyme";

import * as EmployeesContext from "../../context";
import { FeedBackList } from "../FeedbackList";

configure({ adapter: new Adapter() });

describe("FeedbackList Component Test", () => {
  let setValue;
  let container;
  beforeEach(() => {
    setValue = (dataParameter) => {
      jest
        .spyOn(EmployeesContext, "useEmployeeState")
        .mockImplementation(() => dataParameter);
    };
    container = document.createElement("div");
  });

  afterEach(cleanup);

  it("Feedback List Component Renders Correctly", () => {
    const data = [
      {
        form: "view",
        formValue: "",
        feedbackValue: {},
        employee: {
          employee_id: 1,
          firstName: "Yama",
          lastName: "Cruz",
          feedbacklist: [
            {
              feedback_id: 1,
              feedback: "Good Job",
              assign_id: 0,
            },
          ],
        },
      },
    ];
    setValue(data);
    render(<FeedBackList />, container);
  });

  it("Check No Feedback Displayed", () => {
    const data = [
      {
        form: "view",
        formValue: "",
        feedbackValue: {},
        employee: {
          employee_id: 1,
          firstName: "Yama",
          lastName: "Cruz",
          feedbacklist: [
            {
              feedback_id: null,
              feedback: null,
              assign_id: null,
            },
          ],
        },
      },
    ];
    setValue(data);
    const wrapper = mount(<FeedBackList />);
    const h5 = wrapper.find("h5").text();
    expect(h5).toEqual("No Feedback");
  });

  it("Check FeedBack Displayed with Update Button", () => {
    const data = [
      {
        form: "view",
        formValue: "",
        feedbackValue: {},
        employee: {
          employee_id: 1,
          firstName: "Yama",
          lastName: "Cruz",
          feedbacklist: [
            {
              feedback_id: 1,
              feedback: "Good Job",
              assign_id: 0,
            },
          ],
        },
      },
    ];
    setValue(data);
    const wrapper = mount(<FeedBackList />);
    const updateButton = wrapper.find("li button").text();
    const feedback = wrapper.find("li p").text();
    expect(updateButton).toEqual("Update");
    expect(feedback).toEqual("Good Job");
  });

  it("Check Feedbacklist Component Test to Match Snapshot Display", () => {
    const data = [
      {
        form: "view",
        formValue: "",
        feedbackValue: {},
        employee: {
          employee_id: 1,
          firstName: "Yama",
          lastName: "Cruz",
          feedbacklist: [
            {
              feedback_id: 1,
              feedback: "Good Job",
              assign_id: 0,
            },
          ],
        },
      },
    ];
    setValue(data);
    const Snapshot = renderer.create(<FeedBackList />).toJSON();
    expect(Snapshot).toMatchSnapshot();
  });
});
