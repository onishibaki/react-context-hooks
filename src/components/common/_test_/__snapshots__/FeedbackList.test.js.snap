// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeedbackList Component Test Check Feedbacklist Component Test to Match Snapshot Display 1`] = `
<div
  className="sc-AxheI dgwBHG feedback-list-container"
>
  <div
    className="sc-Axmtr hrvXcZ"
  >
    <span
      className="sc-fzpans bghwOc"
      onClick={[Function]}
    >
      X
    </span>
    <button
      className="sc-AxmLO guKNKF"
      onClick={[Function]}
    >
      Add Feedback
    </button>
    <button
      className="sc-AxmLO guKNKF"
      onClick={[Function]}
    >
      Assign Feedback
    </button>
    <ul
      className="sc-fzozJi fvsQND"
    >
      <li>
        <p
          className="sc-fzoLsD fpiVWW"
        >
          Good Job
        </p>
        <button
          className="sc-AxmLO guKNKF"
          onClick={[Function]}
        >
          Update
        </button>
      </li>
    </ul>
  </div>
</div>
`;
